chore: Add CLAUDE.md and Claude Code project context (#765)
kyma-project/runtime-watcher: CLAUDE.md

Reviewer: I think this file is too big to start with. The [CLAUDE.md in `kyma-project/lifecycle-manager`](https://github.com/kyma-project/lifecycle-manager/blob/67ad7938f5620b4e621c3524263db93c152ef6f7/CLAUDE.md?plain=1#L1) Has just 66 lines, compared to 200 here.
I would simplify this file by removing (almost) all details on make targets, and implementation details like events API: Claude can infer that from sources.

# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## What this repo is

> runtime-watcher is a **two-component system** that reduces Lifecycle Manager's reconciliation load by forwarding only meaningful resource changes from SKR clusters back to KCP. Instead of periodic polling, LKM reacts to real events.

Reviewer: ```suggestion
runtime-watcher is a **two-component system** that reduces Lifecycle Manager's reconciliation load by forwarding only meaningful resource changes from SKR clusters back to KCP. Instead of periodic polling, KLM reacts to real events.
```

```
SKR cluster                             KCP cluster
┌───────────────────────────┐           ┌────────────────────────────────────┐
│  runtime-watcher          │  mTLS     │  listener (library)                │
│  (ValidatingWebhook)      │──────────►│  SKREventListener HTTP server      │
│  watches configured       │           │  emits GenericEvent on channel     │
│  resources, detects       │           │  → controller-runtime              │
│  spec/status changes      │           │    reconcile queue                 │
│  always allows admission  │           │                                    │
└───────────────────────────┘           └────────────────────────────────────┘
```

The webhook **always allows** admission requests (`Allowed: true`) — it uses the admission flow purely as a change-detection mechanism, not for enforcement.

> 

Reviewer: I would add here "Glossary" section that would explain the meaning of the following acronyms: KCP, SKR, KLM (and others, if used in this document).
These are obvious to us but not necessarily to Claude Agent with limited context window.
**Edit**: Even better idea is to explain every acronym inline, in the first place it's being used, like it's done in `lifecycle-manager`: https://github.com/kyma-project/lifecycle-manager/blob/67ad7938f5620b4e621c3524263db93c152ef6f7/CLAUDE.md?plain=1#L3

## Repo structure — three Go modules

| Directory | Module | Role |
|---|---|---|

> | `listener/` | `github.com/kyma-project/runtime-watcher/listener` | **Library** consumed by KCP operators (Lifecycle Manager) |

Reviewer: ```suggestion
| `listener/` | `github.com/kyma-project/runtime-watcher/listener` | Go module, a **Library** used by KCP operators (Lifecycle Manager) |
```

> | `runtime-watcher/` | `github.com/kyma-project/runtime-watcher/skr` | **Binary** deployed as a ValidatingWebhook on each SKR |

Reviewer: ```suggestion
| `runtime-watcher/` | `github.com/kyma-project/runtime-watcher/skr` | Go project for creating a **Binary** deployed as a ValidatingWebhook on each SKR cluster |
```

| `make run` | Run webhook locally |
| `make docker-build` / `make docker-push` | Container image lifecycle |

### `listener/`

| Target | What it does |
|---|---|
| `make test` | `fmt` + `vet` + unit tests with race detector (`GOFIPS140=v1.0.0`) |
| `make lint` | golangci-lint |
| `make resolve` | `go mod tidy` |
| `make build-verbose` | Build with verbose output |

### Running a single test

```sh
# listener
cd listener && GOFIPS140=v1.0.0 go test -run TestFoo ./pkg/...

# runtime-watcher
cd runtime-watcher && KUBEBUILDER_ASSETS=$(../bin/setup-envtest use 1.32.0 -p path) \
  GOFIPS140=v1.0.0 go test -run TestFoo `go list ./... | grep -v /tests/`
```

### E2E tests (`runtime-watcher/tests/e2e/`)

E2E tests require real KCP and SKR clusters. Set the following env vars before running:

```sh
export KCP_KUBECONFIG=<path-to-kcp-kubeconfig>
export SKR_KUBECONFIG=<path-to-skr-kubeconfig>
```

| Target | What it does |
|---|---|
| `make test` | Run all e2e suites (watcher-enqueue + watcher-metrics) |
| `make watcher-enqueue` | Run "Enqueue Event from Watcher" suite (timeout: 20m) |
| `make watcher-metrics` | Run "Watcher Metrics" suite (timeout: 20m) |

## How it works

> 1. KLM deploys the webhook as a `ValidatingWebhookConfiguration` + Deployment on each SKR (failure policy: `Ignore`).

Reviewer: ```suggestion
1. KLM deploys the webhook as a `ValidatingWebhookConfiguration` + Deployment on each SKR cluster (failure policy: `Ignore`).
```

2. When a watched resource changes on SKR, Kubernetes calls the webhook via admission.
3. For UPDATE operations, the webhook compares old vs new `.spec` or `.status` and only fires if there's an actual change. CREATE and DELETE are always forwarded.
4. The webhook sends an HTTP POST to KCP (`/<KCP_CONTRACT>/<moduleName>/event`) over mTLS using per-SKR client certificates.
5. On KCP, the Istio Gateway terminates TLS, VirtualServices route to the listener, and the listener puts events on a Go channel consumed by the Kyma controller.
6. Module name is extracted from the incoming webhook URL path: `/validate/<moduleName>`.

## Watcher CR

The [Watcher CR](https://github.com/kyma-project/lifecycle-manager/blob/main/api/v1beta2/watcher_types.go) is defined in **Lifecycle Manager** (not this repo). It drives both the `ValidatingWebhookConfiguration` on SKR and the `VirtualService` routing on KCP. Key fields: `spec.resourceToWatch` (GVK), `spec.labelsToWatch` (optional label filter), `spec.field` (`spec` or `status`), `spec.manager` (URL path segment).

## listener — key API surface

> The listener package is consumed by Lifecycle Manager. Its public API:

Reviewer: ```suggestion
The listener package is used by Lifecycle Manager. Its public API:
```

```go
// Create and start the listener (implements controller-runtime Runnable)
l := event.NewSKREventListener(addr, componentName)
mgr.Add(l)

// Receive events in a controller watch
source.Channel(l.ReceivedEvents(), &handler.EnqueueRequestForOwner{...})
```

**`types.WatchEvent`** — the event payload forwarded from SKR:

```go
type WatchEvent struct {
    Watched    ObjectKey               // {Namespace, Name} of the watched resource
    WatchedGvk metav1.GroupVersionKind // GVK of the watched resource
    SkrMeta    SkrMeta                 `json:"-"` // populated by listener from XFCC header
}
```

> The listener serves HTTP on `/v2/{componentName}/event`. It extracts the SKR's runtime ID from the `X-Forwarded-Client-Cert` (XFCC) header injected by Istio — this is how the KCP side identifies which SKR sent the event.

Reviewer: ```suggestion
In the `Lifecycle-Manager` the listener logic is wired to an endpoint that serves HTTP requests on `/v2/{componentName}/event`. It extracts the SKR's runtime ID from the `X-Forwarded-Client-Cert` (XFCC) header injected by Istio — this is how the KCP side identifies which SKR cluster sent the event.
```
